### PR TITLE
add blocklist

### PR DESCRIPTION
**Pull request by piiswrong**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: docs/api/python/gluon/gluon.md [hunk 1]
```diff
@@ -116,6 +116,8 @@ net.hybridize()
 
     Block
     HybridBlock
+    BlockList
+    HybridBlockList
     SymbolBlock
     nn.Sequential
     nn.HybridSequential
```
Comment (szha) [inline docs/api/python/gluon/gluon.md]: Should these two inherit BlockLists?

**Comment by sxjscience** [comment]
Should solve https://github.com/apache/incubator-mxnet/issues/10101, https://github.com/apache/incubator-mxnet/issues/9264

**Comment by chinakook** [comment]
BlockList is necessary.

File: docs/api/python/gluon/gluon.md (hunk 1 above)
Comment (szha) [reply docs/api/python/gluon/gluon.md]: And concurrent layers too https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/contrib/nn/basic_layers.py#L27

File: python/mxnet/gluon/block.py [hunk 2]
```diff
@@ -841,6 +842,140 @@ def hybrid_forward(self, F, x, *args, **kwargs):
         # pylint: disable= invalid-name
         raise NotImplementedError
 
+
+class BlockList(Block):
+    r"""Holds sub-blocks in a list.
+
+    BlockList can be indexed like a regular Python list, but blocks it
+    contains are properly registered, and will be visible by all Block methods::
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/block.py]: what does it mean to be "properly" registered? Can a block be improperly registered?

File: python/mxnet/gluon/block.py [hunk 3]
```diff
@@ -841,6 +842,140 @@ def hybrid_forward(self, F, x, *args, **kwargs):
...
+        offset = len(self)
+        for i, block in enumerate(blocks):
+            self.register_child(block, str(offset + i))
+        return self
+
+
+class HybridBlockList(HybridBlock):
+    r"""Holds hybrid sub-blocks in a list.
+
+    HybridBlockList can be indexed like a regular Python list, but blocks it
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/block.py]: but the* blocks

**Comment by wkcn** [comment]
Hi!
BlockList is useful. Why not merge this PR?

**Comment by chinakook** [comment]
Because Sequential contains BlockList.

**Comment by wkcn** [comment]
@chinakook Thanks.
I don't know it.
`mxnet.gluon.nn.Sequential` is a serial structure.
[python/mxnet/gluon/nn/basic_layers.py#L51](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/nn/basic_layers.py#L51)
```python
class Sequential(Block):
    [...]
    def forward(self, x):
        for block in self._children.values():
            x = block(x)
        return x
```
How to use `Sequential` to implement a parallel structure like multi-column CNN?
[Single-Image Crowd Counting via Multi-Column Convolutional Neural Network [Page:4]](https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/Zhang_Single-Image_Crowd_Counting_CVPR_2016_paper.pdf)

I think `BlockList` is useful.
e.g.
```python
class MCCNN(nn.HybridBlock):
    def __init__(self, num_filter, columns, aggregator, **kwargs):
        super(MCCNN, self).__init__(**kwargs)
        with self.name_scope():
            self.column_blocks = BlockList(self, [get_column(num_filter, kernel_size, dilations) for kernel_size, dilations in columns])
            self.aggregator = get_aggregator(num_filter, aggregator[0], aggregator[1]) 
            self.conv1x1 = nn.Conv2D(channels = 1, kernel_size = (1, 1), activation = 'relu')
    def hybrid_forward(self, F, x):
        column_outputs = [c(x) for c in self.column_blocks]
        x = F.concat(*column_outputs, dim = 1)
        x = self.aggregator(x)
        x = self.conv1x1(x)
        x = F.squeeze(x, axis = 1)
        return x
```

And here is my BlockList implementation.
```python
def BlockList(self, block_list):
    block_names = []
    for block in block_list:
        assert not hasattr(self, block.name)
        setattr(self, block.name, block)
        block_names.append(block.name)
    class _BlockList:
        def __init__(self, block_names, parent):
            self.block_names = block_names
            self.parent = parent
        def __getitem__(self, i):
            return getattr(self.parent, self.block_names[i])
    return _BlockList(block_names, self)
```

**Comment by wkcn** [comment]
@chinakook 
Thank you! It works.

**Comment by sxjscience** [comment]
Should we add BlockList back because most users do not understand the usage of Sequential/HybridSequential as the container? @szha @leezu

**Comment by szha** [comment]
OK. I'm not sure if we'd use the same implementation though. Let's track this as a feature request.

**Comment by wkcn** [comment]
It need  two lines of code to implement blocklist:

block = nn.Sequential()
block.add(*lst)
